File: PythonPlugins/AdminPlus/AdminPlus.py
__title__ = 'AdminPlus'
__author__ = 'Jakkee'
__version__ = '1.8.3'

import clr
clr.AddReferenceByPartialName("Fougerite")
import Fougerite


class AdminPlus:
    def On_PluginInit(self):
        Util.ConsoleLog(__title__ + " by " + __author__ + " Version: " + __version__ + " loaded.", False)
        if not Plugin.IniExists("Settings"):
            Plugin.CreateIni("Settings")
            ini = Plugin.GetIni("Settings")
            ini.AddSetting("Config", "AdminDestroyEnabled", "true")
            ini.AddSetting("Config", "AdminDestroyWeapon", "Uber Hatchet")
            ini.AddSetting("Config", "AdminTPDistance", "30")
            ini.AddSetting("Config", "ModeratorsCanUse", "true")
            ini.AddSetting("Commands", "Player must run duty first?", "true")
            ini.AddSetting("AdminDoors", "76561198135558142", "Jakkee //These are the players who can open any door in the server! (/admin doors) to add/remove yourself")
            ini.Save()
        DataStore.Flush("AdminPlus")
        ini = Plugin.GetIni("Settings")
        DataStore.Add("AdminPlus", "DestroyEnabled", ini.GetSetting("Config", "AdminDestroyEnabled"))
        DataStore.Add("AdminPlus", "Distance", float(ini.GetSetting("Config", "AdminTPDistance")))
        DataStore.Add("AdminPlus", "DestroyWeapon", ini.GetSetting("Config", "AdminDestroyWeapon"))
        DataStore.Add("AdminPlus", "ModeratorsCanUse", ini.GetSetting("Config", "ModeratorsCanUse"))
        DataStore.Add("AdminPlus", "DutyFirst", ini.GetSetting("Commands", "Player must run duty first?"))

    def On_DoorUse(self, Player, DoorUseEvent):
        if self.toggled(Player):
            if DataStore.Get("AdminPlus", "AccessDoors") == "true":
                DoorUseEvent.Open = True

    def isMod(self, id):
        if DataStore.ContainsKey("Moderators", id):
            if DataStore.Get("AdminPlus", "ModeratorsCanUse"):
                return True
            else:
                return False
        else:
            return False

    def On_Command(self, Player, cmd, args):
        if cmd == "admin":
            if Player.Admin or self.isMod(Player.SteamID):
                if len(args) == 0:
                    Player.MessageFrom("AdminPlus", "/tpadmin [name] (Teleports you 30units away from the target)")
                    Player.MessageFrom("AdminPlus", "/tpback (Teleports you back to where you were)")
                    Player.MessageFrom("AdminPlus", "/duty [on / off] (Tells the server if your on duty or not)")
                    Player.MessageFrom("AdminPlus", "/admin [on / off] (Invisible suit)")
                    Player.MessageFrom("AdminPlus", "/admin metal (Spawns metal building parts)")
                    Player.MessageFrom("AdminPlus", "/admin wood (Spawns wood building parts)")
                    Player.MessageFrom("AdminPlus", "/admin uber (Spawns uber items)")
                    Player.MessageFrom("AdminPlus", "/admin kevlar (Spawns kevlar)")
                    Player.MessageFrom("AdminPlus", "/admin clear (Clears your inventory)")
                    Player.MessageFrom("AdminPlus", "/admin doors (Toggles you to open all doors)")
                    Player.MessageFrom("AdminPlus", "/admin weapons (Gives you weapons + Ammo)")
                elif len(args) == 1:
                    if args[0] == "on":
                        if self.dutyfirst(Player):
                            Player.Inventory.RemoveItem(36)
                            Player.Inventory.RemoveItem(37)
                            Player.Inventory.RemoveItem(38)
                            Player.Inventory.RemoveItem(39)
                            Player.Inventory.AddItemTo("Invisible Helmet", 36, 1)
                            Player.Inventory.AddItemTo("Invisible Vest", 37, 1)
                            Player.Inventory.AddItemTo("Invisible Pants", 38, 1)
                            Player.Inventory.AddItemTo("Invisible Boots", 39, 1)
                            Player.MessageFrom("AdminPlus", "You're now Invisible!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "off":
                        if self.dutyfirst(Player):
                            Player.Inventory.RemoveItem(36)
                            Player.Inventory.RemoveItem(37)
                            Player.Inventory.RemoveItem(38)
                            Player.Inventory.RemoveItem(39)
                            Player.MessageFrom("AdminPlus", "You're now visible!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "wood":
                        if self.dutyfirst(Player):
                            Player.Inventory.AddItem("Wood Pillar", 250)
                            Player.Inventory.AddItem("Wood Foundation", 250)
                            Player.Inventory.AddItem("Wood Wall", 250)
                            Player.Inventory.AddItem("Wood Doorway", 250)
                            Player.Inventory.AddItem("Wood Window", 250)
                            Player.Inventory.AddItem("Wood Stairs", 250)
                            Player.Inventory.AddItem("Wood Ramp", 250)
                            Player.Inventory.AddItem("Wood Ceiling", 250)
                            Player.Inventory.AddItem("Metal Door", 15)
                            Player.Inventory.AddItem("Metal Window Bars", 15)
                            Player.MessageFrom("AdminPlus", "Wood building parts spawned!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "weapons":
                        if self.dutyfirst(Player):
                            Player.Inventory.AddItem("Bolt Action Rifle", 1)
                            Player.Inventory.AddItem("M4", 1)
                            Player.Inventory.AddItem("MP5A4", 1)
                            Player.Inventory.AddItem("9mm Pistol", 1)
                            Player.Inventory.AddItem("P250", 1)
                            Player.Inventory.AddItem("Shotgun", 1)
                            Player.Inventory.AddItem("556 Ammo", 250)
                            Player.Inventory.AddItem("9mm Ammo", 250)
                            Player.Inventory.AddItem("Shotgun Shells", 250)
                            Player.MessageFrom("AdminPlus", "Weapons plus Ammo spawned!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "metal":
                        if self.dutyfirst(Player):
                            Player.Inventory.AddItem("Metal Pillar", 250)
                            Player.Inventory.AddItem("Metal Foundation", 250)
                            Player.Inventory.AddItem("Metal Wall", 250)
                            Player.Inventory.AddItem("Metal Doorway", 250)
                            Player.Inventory.AddItem("Metal Window", 250)
                            Player.Inventory.AddItem("Metal Stairs", 250)
                            Player.Inventory.AddItem("Metal Ramp", 250)
                            Player.Inventory.AddItem("Metal Ceiling", 250)
                            Player.Inventory.AddItem("Metal Door", 15)
                            Player.Inventory.AddItem("Metal Window Bars", 15)
                            Player.MessageFrom("AdminPlus", "Metal building parts spawned!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "uber":
                        if self.dutyfirst(Player):
                            Player.Inventory.AddItem("Uber Hatchet", 1)
                            Player.Inventory.AddItem("Uber Hunting Bow", 1)
                            Player.Inventory.AddItem("Arrow", 40)
                            Player.MessageFrom("AdminPlus", "Uber items spawned!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "kevlar":
                        if self.dutyfirst(Player):
                            Player.Inventory.RemoveItem(36)
                            Player.Inventory.RemoveItem(37)
                            Player.Inventory.RemoveItem(38)
                            Player.Inventory.RemoveItem(39)
                            Player.Inventory.AddItemTo("Kevlar Helmet", 36, 1)
                            Player.Inventory.AddItemTo("Kevlar Vest", 37, 1)
                            Player.Inventory.AddItemTo("Kevlar Pants", 38, 1)
                            Player.Inventory.AddItemTo("Kevlar Boots", 39, 1)
                            Player.MessageFrom("AdminPlus", "I hope you have a legitimate reason why need this!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "clear":
                        if self.dutyfirst(Player):
                            Player.Inventory.ClearAll()
                            Player.MessageFrom("AdminPlus", "Inventory cleared!")
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    elif args[0] == "doors":
                        if self.dutyfirst(Player):
                            self.toggle(Player)
                        else:
                            Player.MessageFrom("AdminPlus", "You're not on duty!")
                    else:
                        Player.MessageFrom("AdminPlus", "Unknown command")
                else:
                    Player.MessageFrom("AdminPlus", "usage: /admin")
            else:
                Player.MessageFrom("AdminPlus", "You're not allowed to use this command!")
        elif cmd == "duty":
            if Player.Admin or self.isMod(Player.SteamID):
                if len(args) == 0:
                    Player.MessageFrom("AdminPlus", "Usage: /duty [on/off]")
                elif len(args) == 1:
                    if args[0] == "on":
                        DataStore.Add("OnDuty", Player.SteamID, "on")
                        Server.Broadcast(Player.Name + " is now on duty! Let him/her know if you need anything!")
                    elif args[0] == "off":
                        DataStore.Remove("OnDuty", Player.SteamID)
                        Server.Broadcast(Player.Name + " is now off duty! Please direct questions to another admin!")
                    else:
                        Player.MessageFrom("AdminPlus", "Unknown Command")
                else:
                    Player.MessageFrom("AdminPlus", "Usage: /duty [on/off]")
            else:
                Player.MessageFrom("AdminPlus", "You're not allowed to use that command!")
        elif cmd == "tpadmin":
            if Player.Admin or self.isMod(Player.SteamID):
                if self.dutyfirst(Player):
                    if len(args) == 0:
                        Player.MessageFrom("AdminPlus", "Usage: /tpadmin [Player Name]")
                    else:
                        target = self.CheckV(Player, args)
                        if target is not None:
                            DataStore.Add("SavedLocation", Player.SteamID, Player.Location)
                            distance = DataStore.Get("AdminPlus", "Distance")
                            if distance is None:
                                distance = float(30)
                            vector3 = Util.Infront(target, distance)
                            Player.SafeTeleportTo(vector3)
                            Player.MessageFrom("AdminPlus", "Teleported: " + str(distance) + "m infront of " + target.Name)
                            Player.MessageFrom("AdminPlus", "Use /tpback to go back to your last location")
                        else:
                            return
                else:
                    Player.MessageFrom("AdminPlus", "You're not on duty!")
            else:
                Player.MessageFrom("AdminPlus", "You're not allowed to use that command!")
        elif cmd == "tpback":
            if Player.Admin or self.isMod(Player.SteamID):
                if self.dutyfirst(Player):
                    if len(args) == 0:
                        if DataStore.Get("SavedLocation", Player.SteamID):
                            BLocation = DataStore.Get("SavedLocation", Player.SteamID)
                            Player.TeleportTo(BLocation)
                            DataStore.Remove("SavedLocation", Player.SteamID)
                            Player.MessageFrom("AdminPlus", "Teleported back!")
                        else:
                            Player.MessageFrom("AdminPlus", "You have no last known locations")
                    elif len(args) > 0:
                        Player.MessageFrom("AdminPlus", "Usage: /tpback")
                else:
                    Player.MessageFrom("AdminPlus", "You're not on duty!")
            else:
                Player.MessageFrom("AdminPlus", "You're not allowed to use that command!")

    def On_PlayerDisconnected(self, Player):
        try:
            if DataStore.Get("SavedLocation", Player.SteamID) is not None:
                DataStore.Remove("SavedLocation", Player.SteamID)
            if DataStore.Get("OnDuty", Player.SteamID) is not None:
                DataStore.Remove("OnDuty", Player.SteamID)
                Server.Broadcast(Player.Name + " is now off duty! Please direct questions to another admin!")
        except:
            pass

    def On_EntityHurt(self, HurtEvent):
        try:
            if HurtEvent.Attacker.Admin:
                if DataStore.Get("AdminPlus", "DestroyEnabled") == "true":
                    if HurtEvent.WeaponName == DataStore.Get("AdminPlus", "DestroyWeapon"):
                        if HurtEvent.Entity.Name is not None:
                            HurtEvent.Entity.Destroy()
        except:
            pass

    def dutyfirst(self, Player):
        try:
            if DataStore.Get("AdminPlus", "DutyFirst") == "true":
                if DataStore.Get("OnDuty", Player.SteamID) == "on":
                    return True
                else:
                    return False
            else:
                return True
        except:
            pass

    def toggle(self, Player):
        ini = Plugin.GetIni("Settings")
        if ini.GetSetting("AdminDoors", Player.SteamID) is None:
            ini.AddSetting("AdminDoors", Player.SteamID, Player.Name)
            ini.Save()
            Player.MessageFrom("AdminPlus", "You have been added to the config!")
            Player.MessageFrom("AdminPlus", "You can now open any door")
        else:
            ini.DeleteSetting("AdminDoors", Player.SteamID)
            ini.Save()
            Player.MessageFrom("AdminPlus", "You have been removed from the config!")
            Player.MessageFrom("AdminPlus", "You can now only open your own doors")

    def toggled(self, Player):
        ini = Plugin.GetIni("Settings")
        if ini.GetSetting("AdminDoors", Player.SteamID) is not None:
            return True
        else:
            return False

    def argsToText(self, args):
        text = str.join(" ", args)
        return text

    def GetPlayerName(self, name):
        try:
            name = name.lower()
            for pl in Server.Players:
                if pl.Name.lower() == name:
                    return pl
            return None
        except:
            return None

    def CheckV(self, Player, args):
        count = 0
        if hasattr(args, '__len__') and (not isinstance(args, str)):
            p = self.GetPlayerName(str.join(" ", args))
            if p is not None:
                return p
            for pl in Server.Players:
                for namePart in args:
                    if namePart.lower() in pl.Name.lower():
                        p = pl
                        count += 1
                        continue
        else:
            nargs = str(args).lower()
            p = self.GetPlayerName(nargs)
            if p is not None:
                return p
            for pl in Server.ActivePlayers:
                if nargs in pl.Name.lower():
                    p = pl
                    count += 1
                    continue
        if count == 0:
            Player.Message("Couldn't find " + str.join(" ", args) + "!")
            return None
        elif count == 1 and p is not None:
            return p
        else:
            Player.Message("Found " + str(count) + " players with similar a name. Use a more correct name!")
            return None

File: CSharpPlugins/AdminPlus/AdminPlus.cs
using Fougerite;
using System;
using System.Collections.Generic;
using System.IO;
using RustBuster2016Server;
using UnityEngine;

namespace AdminPlus
{
    public class AdminPlus : Fougerite.Module
    {
        public bool Admin_Destroy;
        public bool Moderators_Use;
        public bool Use_On_Duty;
        public string Admin_Destroy_Weapon;
        public float Admin_TP_Distance;
        private IniParser Settings;
        private List<string> ToggledUsers = new List<string> { "a0e1f8200a5f22f49567f527fd53d022791b1b8f" };
        private List<string> OnDutyUsers = new List<string> { "a0e1f8200a5f22f49567f527fd53d022791b1b8f" };
        private Dictionary<String, Vector3> TeleportData = new Dictionary<string, Vector3> { };

        #region Plugin Info

        public override string Name
        {
            get
            {
                return "AdminPlus";
            }
        }

        public override string Author
        {
            get
            {
                return "Jakkee";
            }
        }

        public override string Description
        {
            get
            {
                return "[All-In-One]Admin tool that makes you Invisible, Spawns kits, Clears inventory, AdminTP/Doors + MORE";
            }
        }

        public override Version Version
        {
            get
            {
                return new Version("2.0.1");
            }
        }

        #endregion

        #region Initialize

        public override void Initialize()
        {
            if (!File.Exists(Path.Combine(ModuleFolder, "Settings.ini")))
            {
                File.Create(Path.Combine(ModuleFolder, "Settings.ini")).Dispose();
                Settings = new IniParser(Path.Combine(ModuleFolder, "Settings.ini"));
                Settings.AddSetting("Config", "Use Admin destroy", "True");
                Settings.AddSetting("Config", "Moderators can use", "True");
                Settings.AddSetting("Config", "Use /duty", "true");
                Settings.AddSetting("Config", "Admin destroy weapon", "Uber Hatchet");
                Settings.AddSetting("Config", "Admin TP distance", "30");
                Settings.Save();
            }
            else
            {
                Settings = new IniParser(Path.Combine(ModuleFolder, "Settings.ini"));
                Admin_Destroy = bool.Parse(Settings.GetSetting("Config", "Use Admin destroy"));
                Moderators_Use = bool.Parse(Settings.GetSetting("Config", "Moderators can use"));
                Use_On_Duty = bool.Parse(Settings.GetSetting("Config", "Use /duty"));
                Admin_Destroy_Weapon = Settings.GetSetting("Config", "Admin destroy weapon");
                Admin_TP_Distance = float.Parse(Settings.GetSetting("Config", "Admin TP distance"));
            }

            Hooks.OnCommand += new Hooks.CommandHandlerDelegate(On_Command);
            Hooks.OnDoorUse += new Hooks.DoorOpenHandlerDelegate(On_DoorUse);
            Hooks.OnPlayerDisconnected += new Hooks.DisconnectionHandlerDelegate(On_PlayerDisconnected);
            Hooks.OnEntityHurt += new Hooks.EntityHurtDelegate(On_EntityHurt);

        }

        #endregion

        #region DeInitialize

        public override void DeInitialize()
        {
            Hooks.OnCommand -= new Hooks.CommandHandlerDelegate(On_Command);
            Hooks.OnDoorUse -= new Hooks.DoorOpenHandlerDelegate(On_DoorUse);
            Hooks.OnPlayerDisconnected -= new Hooks.DisconnectionHandlerDelegate(On_PlayerDisconnected);
            Hooks.OnEntityHurt -= new Hooks.EntityHurtDelegate(On_EntityHurt);
        }

        #endregion

        #region Core

        public void On_Command(Fougerite.Player player, string cmd, string[] args)
        {
            if (cmd == "admin")
            {
                if (Allowed(player))
                {
                    if (args.Length.Equals(0))
                    {
                        player.MessageFrom("AdminPlus", "/tpadmin [name] (Teleports you " + Admin_TP_Distance.ToString() + "m away from the target)");
                        player.MessageFrom("AdminPlus", "/tpback (Teleports you back to where you were)");
                        player.MessageFrom("AdminPlus", "/duty [on / off] (Tells the server if your on duty or not)");
                        player.MessageFrom("AdminPlus", "/admin [on / off] (Invisible suit)");
                        player.MessageFrom("AdminPlus", "/admin metal (Spawns metal building parts)");
                        player.MessageFrom("AdminPlus", "/admin wood (Spawns wood building parts)");
                        player.MessageFrom("AdminPlus", "/admin uber (Spawns uber items)");
                        player.MessageFrom("AdminPlus", "/admin kevlar (Spawns kevlar)");
                        player.MessageFrom("AdminPlus", "/admin clear (Clears your inventory)");
                        player.MessageFrom("AdminPlus", "/admin doors (Toggles you to open all doors)");
                        player.MessageFrom("AdminPlus", "/admin weapons (Gives you weapons + Ammo)");
                    }
                    else if (CheckDuty(player))
                    {
                        if (args[0] == "on")
                        {
                            player.Inventory.RemoveItem(36);
                            player.Inventory.RemoveItem(37);
                            player.Inventory.RemoveItem(38);
                            player.Inventory.RemoveItem(39);
                            player.Inventory.AddItemTo("Invisible Helmet", 36, 1);
                            player.Inventory.AddItemTo("Invisible Vest", 37, 1);
                            player.Inventory.AddItemTo("Invisible Pants", 38, 1);
                            player.Inventory.AddItemTo("Invisible Boots", 39, 1);
                            player.MessageFrom("AdminPlus", "You're now Invisible!");
                        }
                        else if (args[0] == "off")
                        {
                            player.Inventory.RemoveItem(36);
                            player.Inventory.RemoveItem(37);
                            player.Inventory.RemoveItem(38);
                            player.Inventory.RemoveItem(39);
                            player.MessageFrom("AdminPlus", "You're now Visible!");
                        }
                        else if (args[0] == "wood")
                        {
                            player.Inventory.AddItem("Wood Pillar", 250);
                            player.Inventory.AddItem("Wood Foundation", 250);
                            player.Inventory.AddItem("Wood Wall", 250);
                            player.Inventory.AddItem("Wood Doorway", 250);
                            player.Inventory.AddItem("Wood Window", 250);
                            player.Inventory.AddItem("Wood Stairs", 250);
                            player.Inventory.AddItem("Wood Ramp", 250);
                            player.Inventory.AddItem("Wood Ceiling", 250);
                            player.Inventory.AddItem("Metal Door", 15);
                            player.Inventory.AddItem("Metal Window Bars", 15);
                            player.MessageFrom("AdminPlus", "Wood building parts spawned!");
                        }
                        else if (args[0] == "weapons")
                        {
                            player.Inventory.AddItem("Bolt Action Rifle", 1);
                            player.Inventory.AddItem("M4", 1);
                            player.Inventory.AddItem("MP5A4", 1);
                            player.Inventory.AddItem("9mm Pistol", 1);
                            player.Inventory.AddItem("P250", 1);
                            player.Inventory.AddItem("Shotgun", 1);
                            player.Inventory.AddItem("556 Ammo", 250);
                            player.Inventory.AddItem("9mm Ammo", 250);
                            player.Inventory.AddItem("Shotgun Shells", 250);
                            player.MessageFrom("AdminPlus", "Weapons plus Ammo spawned!");
                        }
                        else if (args[0] == "metal")
                        {
                            player.Inventory.AddItem("Metal Pillar", 250);
                            player.Inventory.AddItem("Metal Foundation", 250);
                            player.Inventory.AddItem("Metal Wall", 250);
                            player.Inventory.AddItem("Metal Doorway", 250);
                            player.Inventory.AddItem("Metal Window", 250);
                            player.Inventory.AddItem("Metal Stairs", 250);
                            player.Inventory.AddItem("Metal Ramp", 250);
                            player.Inventory.AddItem("Metal Ceiling", 250);
                            player.Inventory.AddItem("Metal Door", 15);
                            player.Inventory.AddItem("Metal Window Bars", 15);
                            player.MessageFrom("AdminPlus", "Metal building parts spawned!");
                        }
                        else if (args[0] == "uber")
                        {
                            player.Inventory.AddItem("Uber Hatchet", 1);
                            player.Inventory.AddItem("Uber Hunting Bow", 1);
                            player.Inventory.AddItem("Arrow", 40);
                            player.MessageFrom("AdminPlus", "Uber items spawned!");
                        }
                        else if (args[0] == "kevlar")
                        {
                            player.Inventory.RemoveItem(36);
                            player.Inventory.RemoveItem(37);
                            player.Inventory.RemoveItem(38);
                            player.Inventory.RemoveItem(39);
                            player.Inventory.AddItemTo("Kevlar Helmet", 36, 1);
                            player.Inventory.AddItemTo("Kevlar Vest", 37, 1);
                            player.Inventory.AddItemTo("Kevlar Pants", 38, 1);
                            player.Inventory.AddItemTo("Kevlar Boots", 39, 1);
                            player.MessageFrom("AdminPlus", "I hope you have a legitimate reason why need this!");
                        }
                        else if (args[0] == "clear")
                        {
                            player.Inventory.ClearAll();
                            player.MessageFrom("AdminPlus", "Inventory cleared!");
                        }
                        else if (args[0] == "doors")
                        {
                            string hwid = GetHWID(player);
                            if (ToggledUsers.Contains(hwid))
                            {
                                ToggledUsers.Remove(hwid);
                                player.MessageFrom("AdminPlus", "You can now only open your own doors");
                            }
                            else
                            {
                                ToggledUsers.Add(hwid);
                                player.MessageFrom("AdminPlus", "You can now open any door");
                            }
                        }
                        else
                        {
                            player.MessageFrom("AdminPlus", "/tpadmin [name] (Teleports you " + Admin_TP_Distance.ToString() + "m away from the target)");
                            player.MessageFrom("AdminPlus", "/tpback (Teleports you back to where you were)");
                            player.MessageFrom("AdminPlus", "/duty [on / off] (Tells the server if your on duty or not)");
                            player.MessageFrom("AdminPlus", "/admin [on / off] (Invisible suit)");
                            player.MessageFrom("AdminPlus", "/admin metal (Spawns metal building parts)");
                            player.MessageFrom("AdminPlus", "/admin wood (Spawns wood building parts)");
                            player.MessageFrom("AdminPlus", "/admin uber (Spawns uber items)");
                            player.MessageFrom("AdminPlus", "/admin kevlar (Spawns kevlar)");
                            player.MessageFrom("AdminPlus", "/admin clear (Clears your inventory)");
                            player.MessageFrom("AdminPlus", "/admin doors (Toggles you to open all doors)");
                            player.MessageFrom("AdminPlus", "/admin weapons (Gives you weapons + Ammo)");
                        }
                    }
                    else
                    {
                        player.MessageFrom("AdminPlus", "You're not on duty!");
                    }
                }
                else
                {
                    player.MessageFrom("AdminPlus", "You're not allowed to use this command!");
                }
            }
            else if (cmd == "duty")
            {
                if (Allowed(player))
                {
                    if (args.Length.Equals(0))
                    {
                        player.MessageFrom("AdminPlus", "Usage: /duty [on/off]");
                    }
                    else if (args[0] == "on")
                    {
                        string hwid = GetHWID(player);
                        if (OnDutyUsers.Contains(hwid))
                        {
                            player.MessageFrom("AdminPlus", "You are already on duty!");
                        }
                        else
                        {
                            OnDutyUsers.Add(hwid);
                            Fougerite.Server.GetServer().BroadcastFrom("AdminPlus", player.Name + " is now on duty! Let him/her know if you need anything!");
                        }
                    }
                    else if (args[0] == "off")
                    {
                        string hwid = GetHWID(player);
                        if (OnDutyUsers.Contains(hwid))
                        {
                            OnDutyUsers.Remove(hwid);
                            Fougerite.Server.GetServer().BroadcastFrom("AdminPlus", player.Name + " is now off duty! Please direct questions to another admin!");
                        }
                        else
                        {
                            player.MessageFrom("AdminPlus", "You are already off duty!");
                        }
                    }
                    else
                    {
                        player.MessageFrom("AdminPlus", "Usage: /duty [on/off]");
                    }
                }
                else
                {
                    player.MessageFrom("AdminPlus", "You're not allowed to use this command!");
                }
            }
            else if (cmd == "tpadmin")
            {
                if (Allowed(player))
                {
                    if (CheckDuty(player))
                    {
                        if (args.Length.Equals(0))
                        {
                            player.MessageFrom("AdminPlus", "Usage: /tpadmin [Player Name]");
                        }
                        else
                        {
                            Fougerite.Player target = Fougerite.Server.GetServer().FindPlayer(string.Join(" ", args)) ?? null;
                            if (!target.Equals(null))
                            {
                                string hwid = GetHWID(player);
                                if (!TeleportData.ContainsKey(hwid))
                                {
                                    TeleportData.Add(hwid, player.Location);
                                }
                                Vector3 loc = Util.GetUtil().Infront(target, Admin_TP_Distance);
                                player.TeleportTo(loc.x, loc.y, loc.z);
                                player.MessageFrom("AdminPlus", "Teleported: " + Admin_TP_Distance.ToString() + "m infront of " + target.Name);
                                player.MessageFrom("AdminPlus", "Use /tpback to go back to your last location");
                            }
                            else
                            {
                                player.MessageFrom("AdminPlus", "Can not find player: " + string.Join(" ", args));
                            }
                        }
                    }
                    else
                    {
                        player.MessageFrom("AdminPlus", "You're not on duty!");
                    }
                }
                else
                {
                    player.MessageFrom("AdminPlus", "You're not allowed to use this command!");
                }
            }
            else if (cmd == "tpback")
            {
                if (Allowed(player))
                {
                    if (CheckDuty(player))
                    {
                        string hwid = GetHWID(player);
                        if (TeleportData.ContainsKey(hwid))
                        {
                            Vector3 loc = TeleportData[hwid];
                            player.TeleportTo(loc.x, loc.y, loc.z);
                            TeleportData.Remove(hwid);
                            player.MessageFrom("AdminPlus", "Teleported back!");
                        }
                        else
                        {
                            player.MessageFrom("AdminPlus", "Can not find any previous locations!");
                        }
                    }
                    else
                    {
                        player.MessageFrom("AdminPlus", "You're not on duty!");
                    }
                }
                else
                {
                    player.MessageFrom("AdminPlus", "You're not allowed to use this command!");
                }
            }
        }

        public void On_PlayerDisconnected(Fougerite.Player player)
        {
            try
            {
                string hwid = GetHWID(player);
                if (OnDutyUsers.Contains(hwid))
                {
                    Fougerite.Server.GetServer().BroadcastFrom("AdminPlus", player.Name + " is now off duty! Please direct questions to another admin!");
                    OnDutyUsers.Remove(hwid);
                }
            }
            catch
            {
                Fougerite.Logger.LogDebug("AdminPlus encountered an error during On_PlayerDisconnected (This will NOT effect the plugin)");
            }
        }

        public void On_DoorUse(Fougerite.Player player, Fougerite.Events.DoorEvent DoorEvent)
        {
            if (DoorCheck(player))
            {
                DoorEvent.Open = true;
            }
        }

        public void On_EntityHurt(Fougerite.Events.HurtEvent HurtEvent)
        {
            if (Admin_Destroy)
            {
                if (HurtEvent.AttackerIsPlayer && HurtEvent.VictimIsEntity)
                {
                    if (Allowed(HurtEvent.Attacker as Fougerite.Player))
                    {
                        if (HurtEvent.WeaponName == Admin_Destroy_Weapon)
                        {
                            HurtEvent.Entity.Destroy();
                        }
                    }
                }
            }
        }

        #endregion

        #region Helper Methods

        public bool Allowed(Fougerite.Player player)
        {
            if (player.Admin)
            {
                return true;
            }
            else
            {
                if (player.Moderator && Moderators_Use)
                {
                    return true;
                }
                else
                {
                    return false;
                }
            }
        }

        public bool CheckDuty(Fougerite.Player player)
        {
            if (Use_On_Duty)
            {
                if (OnDutyUsers.Contains(GetHWID(player)))
                {
                    return true;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                return true;
            }
        }

        public bool DoorCheck(Fougerite.Player player)
        {
            if (ToggledUsers.Contains(GetHWID(player)))
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public string GetHWID(Fougerite.Player player)
        {
            string hwid = null;
            foreach (API.RustBusterUserAPI obj in API.RustBusterUsersList)
            {
                if (obj.Name == player.Name && obj.SteamID == player.SteamID)
                {
                    hwid = obj.HardwareID;
                    break;
                }
            }
            return hwid;
       }

        #endregion

    }
}